pkg/provider/equinix.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	provider = &EQUINIXProvider{
+		resp.Clust,
+		projectID,
+		"",
+		resp.Metro,
+		map[string]interface{}{},
+	}
+
+	projectManifest := manifest.ProjectManifest{
+		Cluster:  provider.Cluster(),
+		Project:  provider.Project(),
+		Provider: EQUINIX,
+		Region:   provider.Region(),
+		Context:  provider.Context(),
+		Owner:    &manifest.Owner{Email: conf.Email, Endpoint: conf.Endpoint},
+	}
+
+	if err := projectManifest.Configure(); err != nil {
+		return nil, err
+	}
+
+	provider.bucket = projectManifest.Bucket
+	return provider, nil
+}
+
+func equinixFromManifest(man *manifest.Manifest) (*EQUINIXProvider, error) {
+	return &EQUINIXProvider{man.Cluster, man.Project, man.Bucket, man.Region, man.Context}, nil
+}
+
+func (equinix *EQUINIXProvider) CreateBackend(prefix string, ctx map[string]interface{}) (string, error) {
+
+	ctx["Region"] = equinix.Region()
+	ctx["Bucket"] = equinix.Bucket()
+	ctx["Prefix"] = prefix
+	ctx["ClusterCreated"] = false
+	ctx["__CLUSTER__"] = equinix.Cluster()
+	if cluster, ok := ctx["cluster"]; ok {
+		ctx["Cluster"] = cluster
+		ctx["ClusterCreated"] = true
+	} else {
+		ctx["Cluster"] = fmt.Sprintf(`"%s"`, equinix.Cluster())
+	}
+
+	utils.WriteFile(filepath.Join(equinix.Bucket(), ".gitignore"), []byte("!/*"))
+	utils.WriteFile(filepath.Join(equinix.Bucket(), ".gitattributes"), []byte("/* filter=plural-crypt diff=plural-crypt"))
+
+	return template.RenderString(equinixBackendTemplate, ctx)
+}
+
+func (equinix *EQUINIXProvider) KubeConfig() error {
+	// TODO: deal with current configured KUBECONFIG
+	// TODO: deal with KUBECONFIG env var if it is set, as then the output KUBECONFIG file will be used
+	if utils.InKubernetes() {
+		return nil
+	}
+
+	usr, _ := user.Current()
+
+	input, err := ioutil.ReadFile(filepath.Join(usr.HomeDir, ".kube/config"))
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filepath.Join(usr.HomeDir, ".kube/config-plural-backup"), input, 0644)
+	if err != nil {
+		return err
+	}
+
+	currentDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	kubeConfigFiles := []string{
+		filepath.Join(usr.HomeDir, ".kube/config-plural-backup"),
+		// TODO: make file path more robust
+		filepath.Join(currentDir, "../bootstrap/terraform/kube_config_cluster.yaml"),
+	}
+	kubeconfigs := []*clientcmdapi.Config{}
+
+	for _, filename := range kubeConfigFiles {
+		if len(filename) == 0 {
+			// no work to do
+			continue
+		}
+
+		config, err := clientcmd.LoadFromFile(filename)
+
+		if err != nil {
+			return err
+		}
+
+		kubeconfigs = append(kubeconfigs, config)
+	}
+
+	// first merge all of our maps
+	mapConfig := clientcmdapi.NewConfig()
+
+	for _, kubeconfig := range kubeconfigs {
+		mergo.Merge(mapConfig, kubeconfig, mergo.WithOverride)
+	}
+
+	// merge all of the struct values
+	nonMapConfig := clientcmdapi.NewConfig()
+	for i := range kubeconfigs {
+		kubeconfig := kubeconfigs[i]
+		mergo.Merge(nonMapConfig, kubeconfig, mergo.WithOverride)
```

Comment:
I'm not quite following why this is different to what was done on line 168, should we have not set the mergo.WithOverride opt this time or somethign?

Reply:
I think the comment on line 178 and 179 explain why this is necessary. I actually based this off the code from Kubernetes client-go which I was running into an issue with.
https://github.com/kubernetes/client-go/issues/901